src/pubsub/message/sign.js

```diff
@@ -0,0 +1,92 @@
+'use strict'
+
+const PeerId = require('peer-id')
+const { Message } = require('./index')
+const uint8ArrayConcat = require('uint8arrays/concat')
+const uint8ArrayFromString = require('uint8arrays/from-string')
+const SignPrefix = uint8ArrayFromString('libp2p-pubsub:')
+
+/**
+ * Signs the provided message with the given `peerId`
+ *
+ * @param {PeerId} peerId
+ * @param {Message} message
+ * @returns {Promise<Message>}
+ */
+async function signMessage (peerId, message) {
+  // Get the message in bytes, and prepend with the pubsub prefix
+  const bytes = uint8ArrayConcat([
+    SignPrefix,
+    Message.encode(message)
+  ])
+
+  const signature = await peerId.privKey.sign(bytes)
+
+  return {
+    ...message,
+    signature: signature,
+    key: peerId.pubKey.bytes
+  }
+}
+
+/**
+ * Verifies the signature of the given message
+ * @param {InMessage} message
+ * @returns {Promise<Boolean>}
+ */
+async function verifySignature (message) {
+  // Get message sans the signature
+  const baseMessage = { ...message }
+  delete baseMessage.signature
+  delete baseMessage.key
+  if (typeof baseMessage.from === 'string') {
+    baseMessage.from = PeerId.createFromB58String(baseMessage.from).toBytes()
+  }
+  const bytes = uint8ArrayConcat([
+    SignPrefix,
+    Message.encode(baseMessage)
+  ])
+
+  // Get the public key
+  const pubKey = await messagePublicKey(message)
+
+  // verify the base message
+  return pubKey.verify(bytes, message.signature)
+}
+
+/**
+ * Returns the PublicKey associated with the given message.
+ * If no, valid PublicKey can be retrieved an error will be returned.
+ *
+ * @param {InMessage} message
+ * @returns {Promise<PublicKey>}
+ */
+async function messagePublicKey (message) {
+  // should be available in the from property of the message (peer id)
+  let from
+  if (typeof message.from === 'string') {
+    from = PeerId.createFromB58String(message.from)
+  } else {
+    from = PeerId.createFromBytes(message.from)
+  }
```

Comment:
`PeerId.createFromCID` will handle both these cases.

```suggestion
  const from = PeerId.createFromCID(message.from)
```